fiat-ldap/src/main/java/com/netflix/spinnaker/fiat/roles/ldap/LdapUserRolesProvider.java

```diff
@@ -100,22 +107,56 @@ public List<Role> loadRoles(ExternalUser user) {
         .collect(Collectors.toList());
   }
 
-  private class UserGroupMapper implements AttributesMapper<List<Pair<String, Role>>> {
+  class UserGroupMapper implements AttributesMapper<List<Pair<String, Role>>> {
+
+    @Override
     public List<Pair<String, Role>> mapFromAttributes(Attributes attrs) throws NamingException {
       String group = attrs.get(configProps.getGroupRoleAttributes()).get().toString();
       Role role = new Role(group).setSource(Role.Source.LDAP);
-      List<Pair<String, Role>> member = new ArrayList<>();
-      for (NamingEnumeration<?> members = attrs.get(configProps.getGroupUserAttributes()).getAll();
-          members.hasMore(); ) {
+      List<Pair<String, Role>> members = new ArrayList<>();
+      log.debug(
+          "Parsing out members of the LDAP group {} using attributes {}",
+          group,
+          configProps.getGroupUserAttributes());
+      for (NamingEnumeration<?> groupMembers =
+              attrs.get(configProps.getGroupUserAttributes()).getAll();
+          groupMembers.hasMore(); ) {
         try {
-          String user =
-              String.valueOf(configProps.getUserDnPattern().parse(members.next().toString())[0]);
-          member.add(Pair.of(user, role));
+          String user = getUser(groupMembers.next().toString());
+          members.add(Pair.of(user, role));
+          log.trace("Found user {} for the ldap group {}", user, group);
         } catch (ParseException e) {
           e.printStackTrace();
         }
       }
-      return member;
+      log.debug("Found following member role pairs for the group {}: {}", group, members);
+      return members;
+    }
+
+    private String getUser(String member) throws ParseException {
+      if (configProps.isEnableDnBasedMultiLoad()) {
+        return member.toLowerCase();
+      } else {
+        return String.valueOf(configProps.getUserDnPattern().parse(member)[0]);
+      }
+    }
+  }
+
+  /** Mapper for mapping user distinguished names to their ids */
+  class UserDNMapper implements ContextMapper<Pair<String, String>> {
+
+    @Override
+    public Pair<String, String> mapFromContext(Object ctx) {
+      DirContextAdapter context = (DirContextAdapter) ctx;
+      String userDN =
+          LdapNameBuilder.newInstance(LdapUtils.parseRootDnFromUrl(configProps.getUrl()))
```

Comment:
Thank you for not using the deprecated `DistinguishedName` here! :)

Reply:
Yeah, but the existing code still uses it!!

Reply:
Yeah, that's why I even noticed this change 😁